### PR TITLE
Fixed #29789 -- Added support for nested relations to FilteredRelation.

### PR DESCRIPTION
**Pull request by felixxm**
#12333 with final edits.

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/filtered_relation/tests.py [hunk 1]
```diff
@@ -424,3 +551,123 @@ def test_aggregate(self):
...
+                book=cls.book2,
+                sale_date=cls.sales_date2,
+                currency=cls.eur,
+                sales=100.00,
+                seller=cls.seller2,
+            ),
+        ])
+
+    def test_aggregate(self):
+        tests = [
```
Comment (ferrants) [inline tests/filtered_relation/tests.py:646]: 👍 Nice consolidation here

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -1419,14 +1419,30 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
     def add_filtered_relation(self, filtered_relation, alias):
         filtered_relation.alias = alias
         lookups = dict(get_children_from_q(filtered_relation.condition))
-        for lookup in chain((filtered_relation.relation_name,), lookups):
-            lookup_parts, field_parts, _ = self.solve_lookup_type(lookup)
+        relation_lookup_parts, relation_field_parts, _ = self.solve_lookup_type(filtered_relation.relation_name)
+        if relation_lookup_parts:
+            raise ValueError(
+                "FilteredRelation's relation_name cannot contain lookups "
```
Comment (ferrants) [inline django/db/models/sql/query.py:1425]: 👍

**Comment by ticosax** [comment]
:clap:

**Comment by borismu** [comment]
Hi @ferrants, are there any plans moving it to release? If yes, when could it potentially happen?

It was [merged into master on 9th of July](https://github.com/django/django/blame/master/django/db/models/sql/query.py#L1445) when 3.0 was still in mainstream support. But I don't see it neither [in 3.0](https://github.com/django/django/blame/stable/3.0.x/django/db/models/sql/query.py#L1415) nor [in 3.1](https://github.com/django/django/blame/stable/3.1.x/django/db/models/sql/query.py#L1410). 

Our team uses this feature. We have our own django 2.2 fork with this feature cherrypicked. Now we plan migrating to 3.1. If this feature would be released we would prefer using official django 3.1 rather than forking own version. This way we could receive patch releases automatically.

@felixxm @ticosax Huge thanks for implementing it

**Comment by felixxm** [comment]
@borismu We don't backport new features after an [alpha release](https://docs.djangoproject.com/en/3.1/internals/release-process/#release-cycle). This feature will be available in Django 3.2+.

**Comment by borismu** [comment]
@felixxm Thank you for prompt response!

Looking forward to Django 3.2 then.

**Pull request merged**